### PR TITLE
JPMS fix: split urlconnection into two packages

### PR DESCRIPTION
**Pull request by unverbraucht**
Java modules can have no overlapping exported packages. okhttp and okhttp-urlconnection both exported the package okhttp which resulted in a package split. This PR splits okhttp-urlconnection into two packages as proposed in #4184 : okhttp-java-net-authenticator and okhttp-java-net-cookiejar.

CLA has been signed. Unit tests are all green here. connected tests were all skipped for some reasong for a 12L emuator and a 12.0 device, not sure what happened.

### CODE REVIEW

**Comment by yschimke** [comment]
Looks good to me based on previous discussion, but will defer to @swankjesse for the 2 packages.

**Comment by yschimke** [comment]
We exclude some tests in awkward configurations, even locally.

But you can add the android label to see tests run on emulator in CI.

File: okhttp-java-net-authenticator/src/main/kotlin/okhttp3/java_net/authenticator/JavaNetAuthenticator.kt [hunk 1]
```diff
@@ -13,11 +13,14 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package okhttp3
+package okhttp3.java_net.authenticator
```
[the remark this answers is not in the document]
Comment (unverbraucht) [reply okhttp-java-net-authenticator/src/main/kotlin/okhttp3/java_net/authenticator/JavaNetAuthenticator.kt:16]: Sure, I also think it looks better. Should I move forward with renaming the packages?

**Comment by unverbraucht** [comment]
I fixed a reference to the old subproject names in the native image tests that caused tests to fail.

File: okhttp-java-net-authenticator/src/main/kotlin/okhttp3/java_net/authenticator/JavaNetAuthenticator.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (yschimke) [reply okhttp-java-net-authenticator/src/main/kotlin/okhttp3/java_net/authenticator/JavaNetAuthenticator.kt:16]: Yep. But still waiting on Jesse, just back from vacation I think.

[the remark this answers is not in the document]
Comment (swankjesse) [reply okhttp-java-net-authenticator/src/main/kotlin/okhttp3/java_net/authenticator/JavaNetAuthenticator.kt:16]: Yep. `package okhttp3.javanet.authenticator`

File: okhttp-java-net-authenticator/README.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
+OkHttp URLConnection
+====================
```
Comment (swankjesse) [inline okhttp-java-net-authenticator/README.md:2]:
```suggestion
OkHttp java.net.Authenticator
=============================
```

File: okhttp-java-net-authenticator/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  kotlin("jvm")
+  id("org.jetbrains.dokka")
+  id("com.vanniktech.maven.publish.base")
+  id("binary-compatibility-validator")
+}
+
+project.applyOsgi(
+  "Fragment-Host: com.squareup.okhttp3; bundle-version=\"\${range;[==,+);\${version_cleanup;${projects.okhttp.version}}}\"",
+  "Automatic-Module-Name: okhttp3.java_net.authenticator",
+  "Bundle-SymbolicName: com.squareup.okhttp3.java_net.authenticator",
```
Comment (swankjesse) [inline okhttp-java-net-authenticator/build.gradle.kts:14]:
```suggestion
  "Bundle-SymbolicName: com.squareup.okhttp3.javanet.authenticator",
```

File: okhttp-java-net-cookiejar/README.md [hunk 4]
```diff
@@ -0,0 +1,10 @@
+OkHttp URLConnection
+====================
```
Comment (swankjesse) [inline okhttp-java-net-cookiejar/README.md:2]:
```suggestion
OkHttp java.net.CookieJar
=========================
```

File: okhttp-java-net-cookiejar/README.md
Comment (swankjesse) [inline okhttp-java-net-cookiejar/README.md:1]: rename to `okhttp-javanet-cookiehandler`

File: okhttp-urlconnection/README.md [hunk 5]
```diff
@@ -1,10 +0,0 @@
-OkHttp URLConnection
```
Comment (swankjesse) [inline okhttp-urlconnection/README.md:1]: Could you revert these deletes? I’d like to continue to build & ship this old artifact forever. Breaking users sucks.

File: okhttp/build.gradle.kts [hunk 6]
```diff
@@ -110,7 +110,8 @@ kotlin {
       dependencies {
         dependsOn(commonTest)
         implementation(projects.okhttpTls)
-        implementation(projects.okhttpUrlconnection)
```
Comment (swankjesse) [inline okhttp/build.gradle.kts:113]: Restore this too?

**Review by swankjesse — APPROVED** [review]
I can get behind this, especially with OkHttp 5.0.

But I’d like to do this by adding a new thing, not by renaming an existing thing. Renaming something that already exists is how we lose our users’ trust. It breaks them! It’s especially destructive for users in large repos where it’s difficult to upgrade everything atomically.

**Comment by swankjesse** [comment]
I created a new PR with these changes and a few small naming changes.
https://github.com/square/okhttp/pull/8074

I also removed the javanet.authenticator submodule; it no longer has utility! I added a deprecation note on the original class to explain why.

Thanks for driving this!

**Pull request closed without merging**